### PR TITLE
Add GPU selection to installer

### PR DESCRIPTION
**Pull request by webfiltered**
### GPU installation options

- Adds GPU selection screen
- If a supported GPU is detected, it is pre-selected
- Allows user to pick whatever they wish
- Installs the appropriate `pytorch` for the selection
- Includes support for CPU mode
- Adds flag that auto-starts in CPU mode
- Stores selected GPU in config.json for future ref / debugging

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -96,6 +100,8 @@ interface HardwareValidation {
  */
 export async function validateHardware(): Promise<HardwareValidation> {
   try {
+    if (DesktopConfig.devCpuMode) return { isValid: true, gpu: 'cpu' };
```
Comment (huchenlei) [inline src/utils.ts:103]: Let's change the gpu wording in the PR to `device` instead as the value can be `cpu` as well.

**Review by huchenlei — CHANGES_REQUESTED** [review]

**Review by robinjhuang — APPROVED** [review]

File: src/preload.ts [hunk 2]
```diff
@@ -229,6 +235,30 @@ const electronAPI = {
   showContextMenu: (options?: ElectronContextMenuOptions): void => {
     return ipcRenderer.send(IPC_CHANNELS.SHOW_CONTEXT_MENU, options);
   },
+  Config: {
+    /**
+     * Controls whether ComfyUI core will start up with the `--cpu` flag.
+     * @param settings An key/value pair object containing all settings to update
+     */
+    setCpuMode: async (enabled: boolean): Promise<void> => {
```
Comment (huchenlei) [inline src/preload.ts:243]: Where is this `setCpuMode` function get called? I cannot find its usage in the frontend PR: https://github.com/Comfy-Org/ComfyUI_frontend/pull/1888/files

File: src/preload.ts [hunk 3]
```diff
@@ -229,6 +235,30 @@ const electronAPI = {
...
+      await ipcRenderer.invoke(IPC_CHANNELS.SET_CPU_MODE, enabled);
+    },
+    /**
+     * Finds the name of the last detected GPU type.  Detection only runs during installation.
+     * @returns The last GPU detected by `validateHardware` - runs during installation
+     */
+    getGpu: async (): Promise<TorchDeviceType | undefined> => {
```
Comment (huchenlei) [inline src/preload.ts:250]: Let's rename this to`getDevice`.

File: src/store/index.ts [hunk 4]
```diff
@@ -16,4 +18,12 @@ export type DesktopSettings = {
    * in the yaml config.
    */
   installState?: 'started' | 'installed' | 'upgraded';
+  devCpuMode?: boolean;
```
Comment (huchenlei) [inline src/store/index.ts:21]: Do we really need this flag? Can we compute this from `selectedDevice`? Or is there a difference between user cpu mode and dev cpu mode?

File: src/utils.ts [hunk 5]
```diff
@@ -88,6 +90,8 @@ const execAsync = promisify(exec);
 
 interface HardwareValidation {
   isValid: boolean;
+  /** The detected GPU (not guaranteed to be valid - check isValid) */
+  gpu?: TorchDeviceType;
```
Comment (huchenlei) [inline src/utils.ts:94]: Rename `gpu` to `device`.

File: src/store/index.ts (hunk 4 above)
Comment (huchenlei) [reply src/store/index.ts:21]: If we need to pass extra arg to ComfyUI main.py, we can always set `launchArgs` in the user settings json.

**Review by huchenlei — CHANGES_REQUESTED** [review]

**Pull request merged**